### PR TITLE
Add support for Address Sanitizer in Windows (Dynamic Linkage only)

### PR DESCRIPTION
**Pull request by Maximus-**
This enables Address Sanitizer with MSVC on Windows.

When building this statically, there's some duplicate symbol errors that I think you're supposed to just ignore with `/FORCE:MULTIPLE`, but being that this is only a debug feature, I think it's fine to not bother making it work.

As well, for this to run, the clang ASAN runtime libraries need to be in the system DLL search path. I'm not sure what the correct way to do this is. I usually copy them from `C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.28.29910\bin\Hostx64\x64` into the artifacts directory.

(`clang_rt.asan_dbg_dynamic-x86_64.dll` and `clang_rt.asan_dynamic-x86_64.dll`)

#606 :)

### CODE REVIEW

**Comment by nibanks** [comment]
Thanks for doing this! You should add a build flavor for this to the CI (Schannel for now), and we could use it with spinquic runs.

**Comment by Maximus-** [comment]
Sounds good, I'll take a stab at it

File: CMakeLists.txt [hunk 1]
```diff
@@ -277,7 +277,12 @@ if(WIN32)
     endif()
 
     if(QUIC_ENABLE_SANITIZERS)
-        message(STATUS "Sanitizers unsupported on this platform.")
+        # This fails when linking statically, so for today require dynamic linkage
+        if (QUIC_STATIC_LINK_CRT)
```
Comment (nibanks) [inline CMakeLists.txt:281]: I would like to get static CRT working still.

**Comment by nibanks** [comment]
I enabled it in the CI and I see a ton of these, which are failing the build:
```
warning C5059: runtime checks and address sanitizer is not currently supported - disabling runtime checks
```

**Comment by Maximus-** [comment]
I'm not sure exactly what that's about. [This](https://developercommunity.visualstudio.com/t/runtime-checks-and-address-sanitizer-are-disabled/1144217) mentions it's due to /RTC, but I don't see that being enabled anywhere, unless it's the result of /MD? Looking into it...

**Comment by Maximus-** [comment]
Ah, might be the usage of /SDL

**Comment by nibanks** [comment]
> Ah, might be the usage of /SDL

Ok. I was able to get it to build locally with `/analyze` being removed. Just pushed a change. @maximus- would you be willing/able to try to get dynamic CRT support working too?

**Comment by nibanks** [comment]
So, apparently spinquic is just crashing on start, and running under a debugger doesn't help. Ideas?

**Comment by ThadHouse** [comment]
It's because the Asan runtime dlls are not on path, but are only in the compiler folder. So in order to run, those dlls need to be put somewhere where the tools can find them.

From the docs, static crt is just not supported with Asan. Which is also why the above dlls need to be copied.

**Comment by nibanks** [comment]
> It's because the Asan runtime dlls are not on path, but are only in the compiler folder. So in order to run, those dlls need to be put somewhere where the tools can find them.

@thadhouse what do you think the best way forward is then? Maybe a prepare machine step (for test mode?) to find and copy the binaries?

**Comment by ThadHouse** [comment]
Near the end of build.ps1 we do the same thing for the PGO libraries, which have the same issue. We could do the same thing if ASAN is selected.  It just copies the binaries into the artifacts folder, which is needed to run on the test machines as well, since those don't have the asan runtimes anywhere.

**Comment by nibanks** [comment]
> Near the end of build.ps1 we do the same thing for the PGO libraries, which have the same issue. We could do the same thing if ASAN is selected. It just copies the binaries into the artifacts folder, which is needed to run on the test machines as well, since those don't have the asan runtimes anywhere.

I tried that (will push soon) but I immediately get an unhandled exception (no symbols 😦 ):
```
0:000> kn
 # Child-SP          RetAddr           Call Site
00 00000009`9235e888 00007ffa`1ce60d0e ntdll!memset+0x61
01 00000009`9235e890 00007ffa`1ce3d9a7 clang_rt_asan_dbg_dynamic_x86_64!_asan_wrap_GlobalSize+0x4e053
02 00000009`9235e8d0 00007ffa`1ce3d17e clang_rt_asan_dbg_dynamic_x86_64!_asan_wrap_GlobalSize+0x2acec
03 00000009`9235e900 00007ffa`1ce65e62 clang_rt_asan_dbg_dynamic_x86_64!_asan_wrap_GlobalSize+0x2a4c3
04 00000009`9235e950 00007ffa`1caa4299 clang_rt_asan_dbg_dynamic_x86_64!_asan_wrap_GlobalSize+0x531a7
05 00000009`9235e9b0 00007ffa`1ce70a75 ucrtbased!initterm+0x59
06 00000009`9235e9f0 00007ffa`1ce70967 clang_rt_asan_dbg_dynamic_x86_64!_asan_wrap_GlobalSize+0x5ddba
07 00000009`9235ea40 00007ffa`1ce70c95 clang_rt_asan_dbg_dynamic_x86_64!_asan_wrap_GlobalSize+0x5dcac
08 00000009`9235ea80 00007ffa`1ce70e91 clang_rt_asan_dbg_dynamic_x86_64!_asan_wrap_GlobalSize+0x5dfda
09 00000009`9235ead0 00007ffa`a72e9707 clang_rt_asan_dbg_dynamic_x86_64!_asan_wrap_GlobalSize+0x5e1d6
0a 00000009`9235eb00 00007ffa`a7330b86 ntdll!RtlActivateActivationContextUnsafeFast+0x127
0b 00000009`9235eb70 00007ffa`a7330936 ntdll!RtlGetSystemGlobalData+0x326
0c 00000009`9235ecb0 00007ffa`a73309a8 ntdll!RtlGetSystemGlobalData+0xd6
0d 00000009`9235ecf0 00007ffa`a739c042 ntdll!RtlGetSystemGlobalData+0x148
0e 00000009`9235ed30 00007ffa`a7389168 ntdll!LdrInitShimEngineDynamic+0x3392
0f 00000009`9235f0e0 00007ffa`a7335585 ntdll!RtlIsNonEmptyDirectoryReparsePointAllowed+0x1ae88
10 00000009`9235f160 00007ffa`a733542e ntdll!LdrInitializeThunk+0x165
11 00000009`9235f3e0 00000000`00000000 ntdll!LdrInitializeThunk+0xe
```
@maximus- any ideas?

**Comment by ThadHouse** [comment]
You might also need to copy llvm-symbolizer.exe from the same directory.

**Comment by nibanks** [comment]
> You might also need to copy llvm-symbolizer.exe from the same directory.

No help.

**Comment by Maximus-** [comment]
> > Near the end of build.ps1 we do the same thing for the PGO libraries, which have the same issue. We could do the same thing if ASAN is selected. It just copies the binaries into the artifacts folder, which is needed to run on the test machines as well, since those don't have the asan runtimes anywhere.
> 
> I tried that (will push soon) but I immediately get an unhandled exception (no symbols 😦 ):
> 
> ```
> 0:000> kn
>  # Child-SP          RetAddr           Call Site
> 00 00000009`9235e888 00007ffa`1ce60d0e ntdll!memset+0x61
> 01 00000009`9235e890 00007ffa`1ce3d9a7 clang_rt_asan_dbg_dynamic_x86_64!_asan_wrap_GlobalSize+0x4e053
> 02 00000009`9235e8d0 00007ffa`1ce3d17e clang_rt_asan_dbg_dynamic_x86_64!_asan_wrap_GlobalSize+0x2acec
> 03 00000009`9235e900 00007ffa`1ce65e62 clang_rt_asan_dbg_dynamic_x86_64!_asan_wrap_GlobalSize+0x2a4c3
> 04 00000009`9235e950 00007ffa`1caa4299 clang_rt_asan_dbg_dynamic_x86_64!_asan_wrap_GlobalSize+0x531a7
> 05 00000009`9235e9b0 00007ffa`1ce70a75 ucrtbased!initterm+0x59
> 06 00000009`9235e9f0 00007ffa`1ce70967 clang_rt_asan_dbg_dynamic_x86_64!_asan_wrap_GlobalSize+0x5ddba
> 07 00000009`9235ea40 00007ffa`1ce70c95 clang_rt_asan_dbg_dynamic_x86_64!_asan_wrap_GlobalSize+0x5dcac
> 08 00000009`9235ea80 00007ffa`1ce70e91 clang_rt_asan_dbg_dynamic_x86_64!_asan_wrap_GlobalSize+0x5dfda
> 09 00000009`9235ead0 00007ffa`a72e9707 clang_rt_asan_dbg_dynamic_x86_64!_asan_wrap_GlobalSize+0x5e1d6
> 0a 00000009`9235eb00 00007ffa`a7330b86 ntdll!RtlActivateActivationContextUnsafeFast+0x127
> 0b 00000009`9235eb70 00007ffa`a7330936 ntdll!RtlGetSystemGlobalData+0x326
> 0c 00000009`9235ecb0 00007ffa`a73309a8 ntdll!RtlGetSystemGlobalData+0xd6
> 0d 00000009`9235ecf0 00007ffa`a739c042 ntdll!RtlGetSystemGlobalData+0x148
> 0e 00000009`9235ed30 00007ffa`a7389168 ntdll!LdrInitShimEngineDynamic+0x3392
> 0f 00000009`9235f0e0 00007ffa`a7335585 ntdll!RtlIsNonEmptyDirectoryReparsePointAllowed+0x1ae88
> 10 00000009`9235f160 00007ffa`a733542e ntdll!LdrInitializeThunk+0x165
> 11 00000009`9235f3e0 00000000`00000000 ntdll!LdrInitializeThunk+0xe
> ```
> 
> @Maximus- any ideas?

Are the build artifacts available on the pipeline? If so, I'll root-cause locally

**Comment by nibanks** [comment]
@maximus- https://dev.azure.com/ms/msquic/_build/results?buildId=160451&view=artifacts&pathAsName=false&type=publishedArtifacts

image

**Comment by Maximus-** [comment]
Good news, the issue doesn't repro locally. Bad news, the issue doesn't repro locally. 

Building with ASAN did succeed and the ASAN library artifacts are there and correct, which is perfect. Can we try running the pipeline again to see if the ASAN issue repros? We may need to set LLVM_SYMBOLIZER_PATH to get llvm-symbolizer to run. 

As well, I'll take a look into static linkage.

**Comment by Maximus-** [comment]
Oh, which artifact were you running that gave you that stack trace? I've only tested with spinquic.

**Comment by nibanks** [comment]
> Oh, which artifact were you running that gave you that stack trace? I've only tested with spinquic and quicinterop.

It repro'ed locally for me with spinquic.

**Comment by ThadHouse** [comment]
I bet I know whats happening. We don't upgrade the VS runtime that often on the test systems. Its possible the runtime isn't new enough on the test systems.

**Comment by Maximus-** [comment]
Also, FYI, when running with ASAN in Windbg, ASAN will trigger a lot of Access Violations (on purpose!). You need to `sxi av` or `sxd av` to let ASAN handle them. I suspect that's actually what happened here.

**Comment by nibanks** [comment]
> Also, FYI, when running with ASAN under Windbg, ASAN will trigger a lot of Access Violations (on purpose!). You need to `sxi av` or `sxd av` to let ASAN handle them.

How do I know if I'm bug free if it keeps triggering exceptions? How do I know which exceptions are my bugs?

**Comment by Maximus-** [comment]
ASAN will halt the program if it's an exception it cannot handle. ASAN sets up a large region of memory that's lazily mapped in, so when access violations happen, it checks first if they're in the region that ASAN owns, if not, it'll re-throw the exception as a second-chance, for windbg to handle.

**Comment by nibanks** [comment]
> ASAN will halt the program if it's an exception it cannot handle. ASAN sets up a large region of memory that's lazily mapped in, so when access violations happen, it checks first if they're in the region that ASAN owns, if not, it'll re-throw the exception as a second-chance, for windbg to handle, or for the program to die.

The stack I pasted above is not in our code. There's no way to run this in automation if we get crashes outside our own code. Spinquic automation is designing around running until we get a crash, and it passes if we don't get any. If other random code is throwing exceptions we can't use it in automation.

**Comment by Maximus-** [comment]
Those exceptions will naturally be caught by ASAN at runtime. The only reason they show up under Windbg is because windbg is registered to handle all exceptions first. If you run spinquic outside of windbg, it'll work just as intended. Under windbg, you'll have to `sxd av; g` and it'll then work exactly as intended.

**Comment by nibanks** [comment]
> Under windbg, you'll have to `sxd av; g` and it'll then work exactly as intended.

Ok. That does seem to do it! Thanks! I'll merge main into this to kick the pipeline one more time to see if it's all good and then we can merge.

File: CMakeLists.txt (hunk 1 above)
Comment (nibanks) [reply CMakeLists.txt:281]: We can (try to) add support in a separate PR.

**Review by nibanks — APPROVED** [review]

**Comment by nibanks** [comment]
Looks like it all ran really well.

**Pull request merged**